### PR TITLE
Tlmblgis 3 2

### PR DESCRIPTION
**Pull request by dovek**
Due: 8/30
@lsloan 
@dlhaines 
Please keep in mind, this is only to implement the use of Shiboleth as a form of authentication. I'm still working on the relationship of the tables and use of the User model to create more meaningful data in the tables (i.e. stop using Barb Jensen and the owner table).

### CODE REVIEW

**Comment by dlhaines** [comment]
I don't see an issue with the code itself but...
- There isn't any detail about what the code should be doing so it is hard to evaluate.
- I wonder if there isn't a better way to manage settings.py.

File: Dockerfile [hunk 1]
```diff
@@ -8,8 +8,9 @@ RUN rm /bin/sh && ln -s /bin/bash /bin/sh
 # Install dependencies
 RUN apt-get update && \
 apt-get dist-upgrade -y && \
-apt-get install -y libmysqlclient-dev git python-pip python-dev apache2 apache2-utils libldap2-dev libsasl2-dev nodejs curl
+apt-get install -y libmysqlclient-dev git python-pip python-dev apache2 apache2-utils libldap2-dev libsasl2-dev nodejs curl libssl-dev libffi-dev xmlsec1
```
Comment (lsloan) [inline Dockerfile:12]: Some of these packages have "-dev" in their names.  What does that mean, exactly?  Are they for development purposes _only_?  Will they need to be replaced by something else for production deployment?

File: Dockerfile [hunk 2]
```diff
@@ -8,8 +8,9 @@ RUN rm /bin/sh && ln -s /bin/bash /bin/sh
...
 
+# TODO: Take out the work directory change here. Suggested directory: /usr/apps/Ocellus/
```
Comment (lsloan) [inline Dockerfile:13]: Why not do it now?

File: hacks_mbof/settings.py [hunk 3]
```diff
@@ -186,3 +186,83 @@
...
+SAML2_URL_PATH = '/accounts/'
+# modify to use port request comes
+SAML2_URL_BASE = getenv('DJANGO_SAML2_URL_BASE', 'http://localhost:18000/accounts/')
+
+INSTALLED_APPS += ('djangosaml2',)
+AUTHENTICATION_BACKENDS = (
+    'django.contrib.auth.backends.ModelBackend',
+    'djangosaml2.backends.Saml2Backend',
+)
+LOGIN_URL = '%slogin/' % SAML2_URL_PATH
```
Comment (lsloan) [inline hacks_mbof/settings.py:201]: For a simple string appending, use `SAML2_URL_PATH + '/login/'` instead.  Putting an extra slash before `login` won't hurt and you won't need to worry about whether the value of `SAML2_URL_PATH` ended with one.

File: hacks_mbof/settings.py [hunk 4]
```diff
@@ -186,3 +186,83 @@
...
+SESSION_EXPIRE_AT_BROWSER_CLOSE = True
+
+from os import path
+import saml2
```
Comment (lsloan) [inline hacks_mbof/settings.py:205]: IMNSHO, I prefer imports to happen at the beginning of the program.

File: hacks_mbof/settings.py [hunk 5]
```diff
@@ -186,3 +186,83 @@
...
+BASEDIR = path.dirname(path.abspath(__file__))
+SAML_CONFIG = {
+  'xmlsec_binary': '/usr/bin/xmlsec1',
+  'entityid': '%smetadata/' % SAML2_URL_BASE,
```
Comment (lsloan) [inline hacks_mbof/settings.py:209]: My earlier comment applies here, too.

And there's a few more places, too.  Just look for percent substitution and consider whether it's really necessary or if simple concatenation is easier to read.  Percent substitution mostly makes sense if it's happening in the middle of a string and especially if more than one placeholder is being filled in.

FWIW, I've been using the [`str.format()`](https://docs.python.org/2/library/stdtypes.html#str.format) method more often lately.  See examples of it in [Format String Syntax](https://docs.python.org/2/library/string.html#format-string-syntax).

File: hacks_mbof/settings.py [hunk 6]
```diff
@@ -186,3 +186,83 @@
...
+  # directory with attribute mapping
+  # 'attribute_map_dir': path.join(BASEDIR, 'attribute-maps'),
+  'name': 'Ocellus',
+  # this block states what services we provide
+  'service': {
+      # we are just a lonely SP
+      'sp': {
+          'name': 'Ocellus',
+          'name_id_format': ('urn:oasis:names:tc:SAML:2.0:'
+                             'nameid-format:transient'),
```
Comment (lsloan) [inline hacks_mbof/settings.py:220]: Isn't Python's string continuation syntax kind of cool?  I think so.

File: hacks_mbof/settings.py [hunk 7]
```diff
@@ -186,3 +186,83 @@
...
+          'authn_requests_signed': 'true',
```
Comment (lsloan) [inline hacks_mbof/settings.py:221]: Should this be `True`, not `'true'`?

File: hacks_mbof/settings.py [hunk 8]
```diff
@@ -186,3 +186,83 @@
...
+
+  # set to 1 to output debugging information
+  'debug': 1,
+
+  # certificate
+  'key_file': path.join(BASEDIR, 'saml/ocellus-saml.key'),  'cert_file': path.join(BASEDIR, 'saml/ocellus-saml.pem'),
+  }
+
+ACS_DEFAULT_REDIRECT_URL = getenv('DJANGO_ACS_DEFAULT_REDIRECT', 'http://localhost:18000/')
+LOGIN_REDIRECT_URL = getenv('DJANGO_LOGIN_REDIRECT_URL', 'http://localhost:18000/')
```
Comment (lsloan) [inline hacks_mbof/settings.py:259]: It seems like this URL is used in a lot of places.  Put it in a variable (constant) and document its purpose.

File: hacks_mbof/settings.py [hunk 9]
```diff
@@ -186,3 +186,83 @@
...
+
+SAML_CREATE_UNKNOWN_USER = True
+
+SAML_ATTRIBUTE_MAPPING = {
+    'uid': ('username', ),
+    'mail': ('email', ),
+    'givenName': ('first_name', ),
+    'sn': ('last_name', ),
```
Comment (lsloan) [inline hacks_mbof/settings.py:267]: Is there a place for "full name"?  The user's preferred name, from `displayName`, could go there.

**Pull request merged**